### DISCUSSION TITLE
Allow onSuccess to be async, causing isLoading to wait for resolution

### DISCUSSION THREAD

**Discussion by jtparrett** [TanStack/query#2209] [Ideas] [answered]
As far as I can tell, this isn't a thing currently.

But it would be really handy if onSuccess could accept an async function which will be awaited before isLoading is set to true.

Alternatively a new onAsyncSuccess(terrible name) option might be simpler to implement.

**Comment by TkDodo** [comment, marked as answer]
> But it would be really handy if onSuccess could accept an async function which will be awaited before isLoading is set to true.

you mean before isLoading is set to false? Anyways, what problem are you trying to solve please?

**Reply by jtparrett** [reply]
Yes, isLoading set to false sorry!

It's not for a specific issue, although the case I attempted to use it in and realised it wasn't a thing looked a little like this:
```
const createPost = useMutation(
  (values) => api.post('/create-post', values),
  {
    async onSuccess(){
      await queryClient.refetchQueries(['allPosts']);
    }
  }
)
```

and so you would perform the mutation, and it would await the refetch before createPost.isLoading is set to false.

**Reply by TkDodo** [reply]
Okay, you are talking about useMutation and not useQuery. onSuccess _is_ async for useMutation for this exact reason.

**Reply by jtparrett** [reply]
Wow. I am so sorry for wasting your time. I’m just not gunna go on the internet for a little while 🤦🏻
